[#16] AWS S3 이미지 업로드

choitaehoon opened the pull request:
- AWS S3 이미지 업로드 구현
- 파일 예외 처리

choitaehoon committed:
[#11] 게시물 작성

- 게시물 작성 구현
- AccountController, LoginController naming 변경

choitaehoon committed:
[#14] AWS S3 환경설정

- AWS의 key값 분리 후 gitignore 적용
- AmazonS3 빈으로 등록

choitaehoon committed:
[#14] AWS S3 환경설정

- javax.annotation-api 종속성 추가

choitaehoon committed:
[#14] AWS S3 환경설정

- PropertySource 제거 후 EnvironmentPostProcessor 생성
- aws 환경설정 예외 처리
- aws 환경 설정, 메시지에 관련된 상수 생성

choitaehoon committed:
[#14] AWS S3 환경설정

- gitignore application-aws.yml로 변경

choitaehoon committed:
[#14] AWS S3 환경설정

- 주석제거

choitaehoon committed:
[#14] AWS S3 환경설정

- EnvironmentPostProcessor 제거
- application.yml 파일안에 aws include

choitaehoon committed:
[#16] Aws S3 이미지 업로드

- AWS S3 이미지 업로드 구현
- 파일 예외 처리

f-lab-dev commented on src/main/java/com/social/instagram/controller/FileController.java:
```diff
@@ -0,0 +1,24 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/upload")
```
`upload`는 행위일까요? 리소스를 표현하는것일까요?

choitaehoon replied:
API 문서를 확인해보니 리소스를 중심으로 표현하는게 맞네요. API 가이드 라인을 파악한 내용은 아래와 같습니다.
1. 리소스를 중심으로 디자인 되며 클라이언트에서 엑세스 할 수 있는 모든 종류의 개체, 데이터 또는 서비스가 리소스에 포함된다.
2. 리소스 URI를 복잡하게 구성하면 나중에 리소스 사이의 관계가 변경 될 경우 유연성이 떨어지므로 컬렉션/항목/컬렉션보다 더 복잡하게 요구하지 않는 것이 좋다.

f-lab-dev commented on src/main/java/com/social/instagram/controller/FileController.java:
```diff
@@ -0,0 +1,24 @@
+@RequestMapping("/upload")
+public class FileController {
+
+    private final AwsS3LocalService awsS3LocalService;
```
이 타입은
1. 구체적인 타입 같습니다.
2. 로컬 환경만을 위한 클래스 같습니다.

그렇다면 S3를 사용하는 것에 대해서는 변경에 유연하지 못한 코드가 될 것 같네요~

choitaehoon replied:
인터페이스를 두어 로컬환경과 프로덕션을 따로 분리해 변경에 유연한 코드를 작성하겠습니다

f-lab-dev commented on src/main/java/com/social/instagram/controller/FileController.java:
```diff
@@ -0,0 +1,24 @@
+    public ResponseEntity<Void> upload(@RequestBody MultipartFile file){
+        awsS3LocalService.upload(file);
+
+        return new ResponseEntity<>(HttpStatus.CREATED);
```
여기는 무조건 동일한 데이터를 리턴하는데요. 그러면 상수로 만들어주면 메모리 할당을 줄여줄 수 있지 않을까요?

choitaehoon replied:
상수로 빼놓은 `ResponseConstants class` 를 사용을 안했네요. 정의한 상수를 사용하겠습니다

f-lab-dev commented on src/main/java/com/social/instagram/exception/ApiExceptionHandler.java:
```diff
@@ -43,4 +44,9 @@
 
+    @ExceptionHandler(AwsS3FileNotUploadException.class)
+    public ResponseEntity<String> handleAwsS3FileNotUploadException() {
+        return RESPONSE_AWS_S3_FILE_NOT_UPLOAD;
```
여기서 왜 예외가 발생했는지 로그는 남겨줘야하지 않을까요?

choitaehoon replied:
서버에 배포하고 에러 발생시 분석할 수 있게 로그를 남기는게 좋을거 같네요

f-lab-dev commented on src/main/java/com/social/instagram/util/aws/AwsS3Connector.java:
```diff
@@ -0,0 +1,48 @@
+
+    private final AmazonS3 amazonS3Client;
+
+    @Value("${aws.s3.bucket}")
```
`@Value`도 생성자 주입이 가능합니다

f-lab-dev commented on src/main/java/com/social/instagram/util/aws/AwsS3Connector.java:
```diff
@@ -0,0 +1,48 @@
+        this.amazonS3Client = amazonS3Client;
+    }
+
+    public void upload(MultipartFile file){
```
1. 이 부분은 그냥 S3 Service 클래스에서 해줘도 되는 동작 같습니다. 
2. 아래의 `getContentTypeAndLength` 메소드는 필드를 사용하지 않아 static으로 선언해줘도 될 것 같습니다.

이 점으로 봤을 때 이 클래스가 존재할 필요가 있을까요?

choitaehoon replied:
`AmazonS3`는 매우 많은 기능들을 제공합니다. 하지만 대부분은 사용하지 않는 메소드들이며 필요한 부분만 `AwsS3Connector`  클래스에 기능들을 정의하고 필요한 메소드만 사용할 수 있도록 감싸는게 더 좋지 않을까라는 생각에 구현했습니다. 또한 사용하지 않는 메소드와 사용하고자 하는 메소드의 naming이 비슷하여 혼동할 수 있는 경우가 있을거라 판단했고 혼동되는 naming이 있을 경우 `AwsS3Connector` 클래스 안에 naming을 자유자재로 정할 수 있어 이러한 문제점도 해결 할 수 있지 않을까 생각합니다

f-lab-dev replied:
그럼 이 클래스는 `amazonS3Client`의 다른 기능을 하지 못하도록 캡슐화시킨다라고 볼 수 있는데 이미 `amazonS3Client`는 public class이기에 여기저기서 사용이 가능합니다. 그렇다면 이미 같은 역할을 하는 클래스를 재정의하는게 아닐까요?

choitaehoon replied:
`amazonS3Client`가 어디서나 접근 가능하기에 재정의 한거라고 볼 수 있겠네여. 그렇다면 `AwsS3LocalService`에 `AmazonS3`를 주입하고 사용하는게 맞다고 생각이 들어 구현하겠습니다

choitaehoon committed:
[#16] Aws S3 이미지 업로드

- FileController API naming 변경
- upload 상수로 반환
- 에러 로그 추가
- AWS S3 인터페이스로 확장

f-lab-dev commented on src/main/java/com/social/instagram/exception/ApiExceptionHandler.java:
```diff
@@ -25,27 +27,32 @@
 
     @ExceptionHandler(UserNotAccountException.class)
     public ResponseEntity<String> handleNotUserDataException() {
+        log.error("입력된 이메일과 일치하는 계정이 없습니다");
```
예외도 같이 로그에 찍어주면 좋을 것 같네요~ 이렇게 문자열만 두면 추후에 디버깅할때 예외와 관련된 정보가 디버깅이 힘듭니다

choitaehoon replied:
아 그렇겠네요. 예외도 같이 로그에 찍겠습니다

f-lab-dev commented on src/main/java/com/social/instagram/util/aws/AwsS3Connector.java:
```diff
@@ -18,17 +19,14 @@
     Ioexception을 처리해줘야 하는데 try-with-resources를 사용하여 자원 해제 처리를 자동으로 종료 할 수 있게 구현했다.
 */
 @Configuration
+@RequiredArgsConstructor
```
`@Value`를 남겨두셨는데요. 이렇게하면 생성자 주입 따로, 필드 주입 따로 돌아가지 않을까요?

choitaehoon replied:
생각을 해보니 `AwsS3Config`에 `bucket`을 빈으로 등록하고 생성자 주입을 받아야 될거 같네요.

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- AwsS3Config 안에 bucket 빈으로 등록
- 예외 에러 로그 추가
- AwsS3Connector에서 AwsS3LocalService로 upload 옮기기

choitaehoon commented on src/main/java/com/social/instagram/service/AwsS3LocalService.java:
```diff
@@ -1,19 +1,42 @@
+            throw new AwsS3FileNotUploadException();
+        }
+
+        log.info("file name : '{}' upload success!", file.getOriginalFilename());
```
Aws에 관한 전반적인 로그를 출력하지 않아 정상적으로 운영되는지 파악하기 힘들어 로그를 남겼습니다. 하지만 이 로그는 횡단 관심사로 분리할 필요가 있습니다. `AwsS3LocalService` 뿐만 아니라 `AwsS3ProductService`에도 재사용 할 수 있으며 메소드가 실행된 후에 정상적으로 작동했다는 로그를 남겨야 하기 때문에 **AOP**를 적용하겠습니다.

f-lab-dev replied:
1. 이거는 upload때마다 엄청난 로그를 출력할텐데 debug로 낮춰서 실서버에서는 찍히지 않도록 하는게 낫지 않을까요?
2. 출력값 로깅은 모두 메세지가 다를텐데 굳이 AOP로 분리해야할까요?

choitaehoon replied:
1. logback의 정책인 `RollingFileAppender`를 이용해서 만료기간이 지나면 로그를 없애는 방식으로 하게 된다면 문제가 없을거 같다고 생각했는데 대규모시스템에선 매우 많은 로그를 출력해서 메모리가 꽉 찰 가능성이 존재하네요. debug로 낮추는게 맞는것 같습니다
2. `AwsS3LocalService`와 `AwsS3ProductService`은 업로드, s3 목록 가져오기등 동일한 기능을 하기 때문에 메시지가 같을거라 생각했고 횡단 관심사를 분리하고자 AOP를 적용해야 겠다는 판단을 했습니다.

f-lab-dev replied:
1. 메모리보다는 로그처리시스템도 컴퓨팅파워를 사용하는데 그 곳에 꽤나 부담이 될겁니다 ㅎㅎ
2. 그러면 범용성보다는 이 클래스 하나만을 위해 AOP를 사용하겠다는 것으로 들리는데요. 그러면 굳이 이 S3클래스만을 위해 명확성을 떨어뜨리면서까지 AOP를 적용할 필요가 있을까요?

예를 들면 태훈님이 퇴사하거나 다른 부서로 간 이후, 모두들 이 AOP의 존재를 까먹는다면 미래에는 필요없는 로그를 찍는 AOP 클래스가 남아버릴 가능성이 있을 것 같은데요~

choitaehoon replied:
제가 AOP에 대해 완벽히 이해하지 못한거 같네요. 토비의 스프링 AOP책과 document를 보면서 학습하겠습니다

f-lab-dev commented on src/main/java/com/social/instagram/config/AwsS3Config.java:
```diff
@@ -31,4 +34,9 @@ public AmazonS3 amazonS3Client() {
     }
 
+    @Bean
+    public String bucket() {
```
그냥 `@Value`로 주입받으면 되는데 굳이 bean으로 선언해주는 이유가 있으실까요?

choitaehoon replied:
`@Value`로 bucket 필드에 값을 주입을 해준 뒤에 bucket을 빈으로 등록한 다음 `AwsS3LocalService`에서 주입 받을 수 있도록 해야 하지 않나요? 그냥 `@Value`로 주입 받게만 한다면 빈을 정의하라는 문구가 나옵니다

choitaehoon replied:
`@Value`로 bucket 필드에 값을 주입 해준 뒤에 bucket을 빈으로 등록한 다음 `AwsS3LocalService`에서 주입 받을 수 있도록 해야 하지 않을까요? 그냥 `@Value`로 주입 받게만 한다면 빈을 정의하라는 문구가 나옵니다

f-lab-dev replied:
생성자 인자에 `@Value`를 붙여보셨나요?

choitaehoon replied:
붙여도 주입이 되지만 AWS 서비스 클래스마다 생성자 인자에 `@Value`를 작성하기 보다 `AwsS3Config`클래스 안에 버킷을 빈으로 등록하고 단순히 `@RequiredArgsConstructor`를 사용하는게 좋을거 같아 `AwsS3Config` 클래스에 빈으로 등록했습니다.

f-lab-dev replied:
그거는 어떻게든 `@RequiredArgsConstructor`를 사용하기 위한 야매로 보입니다~ 스프링에서 지원하는 기능을 우회해서 쓸만큼 `@RequiredArgsConstructor`가 필수적일까요??

choitaehoon replied:
필수는 아니지만 `AwsS3Config`에 빈으로 등록하고 여러 곳에서 생성자 주입을 바로 사용 할 수 있게 하면 간편해지지 않을까 생각했습니다. `AwsS3Config`에 버킷을 `@Value`로 값을 주입하고 빈으로 등록하는 작업이 우회한다고 볼 수 있네요.

f-lab-dev commented on src/main/java/com/social/instagram/exception/ApiExceptionHandler.java:
```diff
@@ -27,32 +27,32 @@
     @ExceptionHandler(UserIdDuplicatedException.class)
     public ResponseEntity<String> handleDuplicateUserIdException() {
-        log.error("중복된 아이디가 존재합니다.");
+        log.error("[{}] 중복된 아이디가 존재합니다.", HttpStatus.BAD_REQUEST);
```
`ExceptionHandler`는 예외객체를 받아올 수 있습니다~ 그걸 받아와서 로그에 찍어주세요. 이것도 디버깅에 도움이 되지 않을 것 같아요~

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- ExceptionHandler를 이용하여 예외 객체 로그 적용
- upload 로그 debug로 낮추기

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- Post Entity에 filePath 추가
- S3 이미지 업로드 후 Post Entity에 filePath, userId 저장

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- PostEntity로 변경하는 메소드 naming 변경

choitaehoon commented on src/main/java/com/social/instagram/controller/FileController.java:
```diff
@@ -0,0 +1,35 @@
+        String userId = sessionService.getUserId();
+        String filePath = awsS3Service.upload(file, userId);
+
+        postService.writePost(Post.changePostEntity(filePath, userId));
```
S3에 이미지를 업로드 하고 `Post`를 데이터베이스에 저장하는 과정에서 중간에 예외가 발생하게 되면 데이터 정합성 이슈가 발생할 가능성이 존재합니다.

`@Transcational`을 이용하여 하나의 작업 단위로 묶으려고 생각을 했지만 이미지 업로드 과정은 원격 서버와의 통신 작업인데 만일 원격 서버와 통신 할 수 없는 상황이 발생하면 db 커넥션을 소유한채 반납하지 않아 사용가능한 커넥션의 수가 줄어들어 어느 순간에 커넥션을 가져가기 위해 다른 요청은 지연될 수 밖에 없는 상황이 발생하는 문제가 발생할 가능성이 있어 `@Transactional`을 적용하지 않았습니다.

이 문제를 해결하기 위해 고민해보겠습니다

choitaehoon commented on src/main/java/com/social/instagram/controller/FileController.java:
```diff
@@ -27,7 +27,7 @@
         String filePath = awsS3Service.upload(file, userId);
 
-        postService.writePost(Post.changePostFilePathAndUserId(filePath, userId));
+        postService.writePost(Post.changePostEntity(filePath, userId));
```
S3에 이미지를 업로드 하고 `Post`를 데이터베이스에 저장하는 과정에서 중간에 예외가 발생하게 되면 데이터 정합성 이슈가 발생할 가능성이 존재합니다.

`@Transcational`을 이용하여 하나의 작업 단위로 묶으려고 생각을 했지만 이미지 업로드 과정은 원격 서버와의 통신 작업인데 만일 원격 서버와 통신 할 수 없는 상황이 발생하면 db 커넥션을 소유한채로 반납하지 않아 사용가능한 커넥션의 수가 줄어들어 어느 순간에 커넥션을 가져가기 위해 다른 요청은 지연될 수 밖에 없는 상황이 발생하는 문제가 발생할 가능성이 있어 `@Transactional`을 적용하지 않았습니다.

데이터 정합성 이슈를 해결하는 방안에 대해 좀 더 고민해보겠습니다.

f-lab-dev replied:
만약 트랜잭션이 실패하면 hook을 걸어서 S3에 있는 이미지를 삭제해주는걸 구현할 수 있지 않을까요?

choitaehoon replied:
그 방법도 좋은거 같네요. 여러개의 방안도 검토 해보고 적용시켜보겠습니다. 이 PR에 대한 코멘트가 길어져 리뷰하실때 불편할거라 생각해 트랜잭션에 관한 부분은 따로 이슈하고 PR을 만들겠습니다

f-lab-dev commented on src/main/java/com/social/instagram/exception/ApiExceptionHandler.java:
```diff
@@ -26,33 +26,33 @@
-    public ResponseEntity<String> handleDuplicateUserIdException() {
-        log.error("[{}] 중복된 아이디가 존재합니다.", HttpStatus.BAD_REQUEST);
+    public ResponseEntity<String> handleDuplicateUserIdException(UserIdDuplicatedException exception) {
+        log.error("[{}] 중복된 아이디가 존재합니다.", exception.toString());
```
로거는 `Exception` 객체를 받는 메소드가 존재합니다. 굳이 toString()으로 찍기보다는 통째로 stacktrace를 찍어서 디버깅이 용이하게 만들어주면 좋을 것 같아요

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- 디버깅에 용이하게 만들고자 ExceptionHandler안에 stacktrace 출력 구현

f-lab-dev commented on src/main/java/com/social/instagram/exception/ApiExceptionHandler.java:
```diff
@@ -27,32 +28,32 @@
     @ExceptionHandler(UserIdDuplicatedException.class)
     public ResponseEntity<String> handleDuplicateUserIdException(UserIdDuplicatedException exception) {
-        log.error("[{}] 중복된 아이디가 존재합니다.", exception.toString());
+        log.error("[{}] {}", "중복된 아이디가 존재합니다.", ExceptionUtils.getStackTrace(exception));
```
이렇게 하면 어떨까요?

```suggestion
        log.error("중복된 아이디가 존재합니다.", exception);
```

choitaehoon replied:
더 깔끔하고 명확해보이네요. 수정하겠습니다

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- Exception Handler error 로그 수정
- AwsS3LocalService @Value 생성자 주입
- AwsS3Config bucket 빈 삭제

choitaehoon committed:
[#16] AWS S3 이미지 업로드

- ExceptionUtils 의존성 삭제

choitaehoon committed:
Merge branch 'develop' into feature/16-aws-upload
